File: retrobuilder/src/main/java/site/business/appslandz/retrobuilder/Service/ApiClient.java
package site.business.appslandz.retrobuilder.Service;

import android.content.Context;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;
import site.business.appslandz.retrobuilder.Interceptors.AuthInterceptor;
import site.business.appslandz.retrobuilder.Interfaces.AuthInitializer;
import timber.log.Timber;

public class ApiClient {

    private final String TAG = ApiClient.class.getSimpleName();

    private static ApiClient ourInstance = null;

    private Context context = null;

    private String baseUrl;

    public static ApiClient getInstance(Context context) {
        if (ourInstance == null)
            ourInstance = new ApiClient(context);
        return ourInstance;
    }

    private ApiClient(Context context) {
        this.context = context;
    }

    public ApiClient setBaseUrl(String baseUrl) {
        this.baseUrl = baseUrl;
        return this;
    }

    public Retrofit getClient() {

        Timber.plant(new Timber.DebugTree());

        // HttpLoggingInterceptor
        HttpLoggingInterceptor httpLoggingInterceptor = new HttpLoggingInterceptor(message -> Timber.i(message));
        httpLoggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);

        OkHttpClient client = new OkHttpClient.Builder()
                .addInterceptor(httpLoggingInterceptor)
                .build();

        return new Retrofit.Builder()
                .baseUrl(this.baseUrl)
                .addConverterFactory(GsonConverterFactory.create())
                .client(client)
                .build();
    }


    public Retrofit getAuthClient(AuthInitializer authInitializer) {

        Timber.plant(new Timber.DebugTree());

        // HttpLoggingInterceptor
        HttpLoggingInterceptor httpLoggingInterceptor = new HttpLoggingInterceptor(message -> Timber.i(message));
        httpLoggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);

        AuthInterceptor authInterceptor = new AuthInterceptor(authInitializer);

        OkHttpClient client = new OkHttpClient.Builder()
                .addInterceptor(httpLoggingInterceptor)
                .addInterceptor(authInterceptor)
                .build();

        return new Retrofit.Builder()
                .baseUrl(this.baseUrl)
                .addConverterFactory(GsonConverterFactory.create())
                .client(client)
                .build();
    }

}

File: retrobuilder/src/main/java/site/business/appslandz/retrobuilder/Service/ApiClient.kt
package site.business.appslandz.retrobuilder.Service

import android.content.Context
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import site.business.appslandz.retrobuilder.Interceptors.AuthInterceptor
import site.business.appslandz.retrobuilder.Interfaces.AuthInitializer
import timber.log.Timber
import timber.log.Timber.DebugTree

class ApiClient private constructor(context: Context) {

    private val TAG = ApiClient::class.java.simpleName
    private var context: Context? = null
    private var baseUrl: String = ""

    init {
        this.context = context
    }

    // HttpLoggingInterceptor
    val client: Retrofit
        get() {
            Timber.plant(DebugTree())
            // HttpLoggingInterceptor
            val httpLoggingInterceptor = HttpLoggingInterceptor(object : HttpLoggingInterceptor.Logger {
                override fun log(message: String) {
                    Timber.i(message)
                }
            })
            httpLoggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY)
            val client = OkHttpClient.Builder()
                    .addInterceptor(httpLoggingInterceptor)
                    .build()
            return Retrofit.Builder()
                    .baseUrl(this.baseUrl)
                    .addConverterFactory(GsonConverterFactory.create())
                    .client(client)
                    .build()
        }

    fun getAuthClient(authInitializer: AuthInitializer?): Retrofit {
        Timber.plant(DebugTree())
        // HttpLoggingInterceptor
        val httpLoggingInterceptor = HttpLoggingInterceptor(object : HttpLoggingInterceptor.Logger {
            override fun log(message: String) {
                Timber.i(message)
            }
        })
        httpLoggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY)
        val authInterceptor = AuthInterceptor(authInitializer!!)
        val client = OkHttpClient.Builder()
                .addInterceptor(httpLoggingInterceptor)
                .addInterceptor(authInterceptor)
                .build()
        return Retrofit.Builder()
                .baseUrl(this.baseUrl)
                .addConverterFactory(GsonConverterFactory.create())
                .client(client)
                .build()
    }

    companion object {

        private var ourInstance: ApiClient? = null

        @kotlin.jvm.JvmStatic
        fun getInstance(context: Context): ApiClient {
            if (ourInstance != null) {
                return ourInstance as ApiClient
            }
            ourInstance = ApiClient(context)
            return ourInstance as ApiClient
        }
    }

    fun setBaseUrl(baseUrl: String): ApiClient {
        this.baseUrl = baseUrl
        return this
    }


}